### PR TITLE
Adding filters to triggers using k8s LabelSelectors

### PR DESCRIPTION
**Pull request by nachocano**
## Proposed Changes

- Adding filters to triggers using k8s Label Selectors. With this approach, we can reuse all the existing machinery for matching labels.
- As k8s LabelSelectors validate label keys and values, we are bypassing that validation for now. In order to match expressions, we will probably need to write our own "FilterSelectorRequirement". In a subsequent PR we will prototype a quick hack using reflection to set unexported fields in a Requirement struct. 

A trigger example looks like this:

```yaml
apiVersion: eventing.knative.dev/v1alpha1
kind: Trigger
metadata:
  name: trigger-test
spec:
  broker: broker-test
  filter:
    headers:
      Ce-Eventtype: '"com.example.someevent"'
      Ce-Source: '"/mycontext/subcontext"'
  subscriber:
    ref:
      apiVersion: serving.knative.dev/v1alpha1
      kind: Service
      name: message-dumper

```

**Release Note**

```release-note
NONE
```

### CODE REVIEW

**Comment by nachocano** [comment]
@Harwayne  I guess the /cc has to be enabled in the repo. Let me know if it makes sense. We can chat more tomorrow.
Thanks!

**Comment by Harwayne** [comment]
All the /commands are prow commands as far as I know. I think using @<name> will inform people (depending on their email settings).

File: t2.yaml [hunk 1]
```diff
@@ -3,12 +3,14 @@ kind: Trigger
...
-  # Our Cloud Event parsing library seems to have a bug and forces to put type and source in double
-  # quotes (as it thinks the actual value is `"foo"`, not `foo`).
-  type: '"com.example.someevent"'
   subscriber:
     ref:
      apiVersion: serving.knative.dev/v1alpha1
      kind: Service
      name: message-dumper
-
+  # Our Cloud Event parsing library seems to have a bug and forces to put type and source in double
```
Comment (Harwayne) [inline t2.yaml]: I think this means that that the values below need to be wrapped in double quotes. Have you tried running this to see what passes through?

File: t2.yaml [hunk 2]
```diff
@@ -3,12 +3,14 @@ kind: Trigger
...
+  # quotes (as it thinks the actual value is `"foo"`, not `foo`).
+  filters:
```
Comment (Harwayne) [inline t2.yaml]: My preference is to put `filters` before `subscriber`.

File: pkg/apis/eventing/v1alpha1/trigger_types.go [hunk 3]
```diff
@@ -58,11 +58,10 @@ type TriggerSpec struct {
...
 	DeprecatedGeneration int64 `json:"generation,omitempty"`
 
-	Broker     string               `json:"broker,omitempty"`
-	Subscriber *SubscriberSpec      `json:"subscriber,omitempty"`
+	Broker     string          `json:"broker,omitempty"`
+	Subscriber *SubscriberSpec `json:"subscriber,omitempty"`
 
-	Type   string `json:"type,omitempty"`
-	Source string `json:"source,omitempty"`
+	Filters *metav1.LabelSelector `json:"filters,omitempty"`
```
Comment (Harwayne) [inline pkg/apis/eventing/v1alpha1/trigger_types.go]: Let's put in another layer of indirection:

```yaml
spec:
  filter:
    headers:
      Ce-Eventtype: com.example.someevent
      ...
```

That gives an easy place to add something like `headerExpressions`, as sibling of `headers`

**Review by Harwayne — COMMENTED** [review]
Looks awesome!

File: pkg/apis/eventing/v1alpha1/trigger_types.go (hunk 3 above)
Comment (nachocano) [reply pkg/apis/eventing/v1alpha1/trigger_types.go]: Done. Having headerExpressions will be sligthly more involved than I thought. There is a Requirement struct that doesn't expose its fields. I'm trying to do a quick prototype using reflection to set those values. For now, I just pushed the headers with exact matching. It works as what you previously had of type and source.

File: t2.yaml (hunk 2 above)
Comment (nachocano) [reply t2.yaml]: Done

File: t2.yaml (hunk 1 above)
Comment (nachocano) [reply t2.yaml]: Done, now it works as your example.

**Comment by nachocano** [comment]
@Harwayne Done the changes. I still have locally my headerExpressions part. I'm doing something ugly with reflection in order to set some unexported fields. Will probably create another PR with that, because you won't like it, and we may need to add our own implementation. I guess this part works as the current one, but with a slightly better definition in the yaml.

**Pull request closed without merging**